core/src/main/scala/scalaz/Name.scala

```diff
@@ -52,13 +52,31 @@ object Name {
   }
 }
 
-object Need {
+/** Synchronized, thread-safe version of `Need` */
+object ThreadSafeNeed {
   def apply[A](a: => A): Need[A] = {
     new Need[A] {
-      private[this] lazy val value0: A = a
-      def value = value0
+      lazy val value: A = a
```

Comment:
I think `private[this]` is important if we want clear reference of the thunk `a: => A`. see https://github.com/scalaz/scalaz/pull/422.
Also, There is another way https://github.com/xuwei-k/scalaz/commit/e52825eb3d